### PR TITLE
Improve FP16 performance of CUDA reflection and replication padding

### LINKED ISSUE #13058: Backward pass over torch.nn.functional.pad is extremely slow with half tensors
## 🐛 Bug

The backward pass over `torch.nn.functional.pad` is more than 300x slower with half tensors compared to fp32 tensors.

## To Reproduce

```
import torch
import time

def exec(x):
  y=torch.nn.functional.pad(x,pad=(47,48,47,48),mode="replicate")
  torch.cuda.synchronize()
  tic=time.time()
  y.sum().backward()
  torch.cuda.synchronize()
  return time.time()-tic

x=torch.rand(4,2048,1,1, requires_grad=True).cuda()
print("fp32: {}".format(exec(x)))
print("fp16: {}".format(exec(x.half())))

```

My output:
```
fp32: 0.14527177810668945
fp16: 50.3135507106781
```

## Expected behavior

At least as fast as fp32.

## Environment
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.2.148

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-16ubuntu3) 7.3.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 396.37
cuDNN version: Probably one of the following:
/usr/local/cuda-9.2/lib64/libcudnn.so.7.2.1
/usr/local/cuda-9.2/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] cuda92                    1.0                           0    pytorch
[conda] pytorch                   0.4.1           py37_cuda9.2.148_cudnn7.1.4_1  [cuda92]  pytorch
[conda] torchfile                 0.1.0                     <pip>
[conda] torchnet                  0.0.4                     <pip>
[conda] torchvision               0.2.1                    py37_1    pytorch

### PR DESCRIPTION
**Pull request by xwang233**
Improve FP16 performance of CUDA reflection and replication padding by replacing `gpuAtomicAdd` with `fastAtomicAdd`.

For the test in https://github.com/pytorch/pytorch/issues/13058, 

before

```
fp32: 0.13376355171203613
fp16: 47.14315724372864
```

after

```
fp32: 0.1379387378692627
fp16: 0.12236762046813965
```

Fix https://github.com/pytorch/pytorch/issues/13058

### CODE REVIEW

File: aten/src/ATen/native/cuda/ReplicationPadding.cu [hunk 1]
```diff
@@ -123,7 +131,13 @@ __global__ void replication_pad_backward_kernel(
   int inputPointY = imin(imax(padT, outputPointY), gradInput.size(2) + padT - 1) - oStartY + iStartY;
 
   scalar_t valueToCopy = gradOutput[batch][plane][outputPointY][outputPointX];
-  gpuAtomicAdd(&gradInput[batch][plane][inputPointY][inputPointX], valueToCopy);
+  fastAtomicAdd(
+    gradInput.data(),
+    idx_2d(batch * channel + plane, inputPointY, inputPointX, height, width),
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ReplicationPadding.cu:136]: packedTensorAccessor supports non-contiguous tensors, yet here you are replacing index computation with index computation that's only valid for contiguous tensors. I don't immediately see if gradInput is guaranteed to be contiguous, probably not.

Comment (xwang233) [reply aten/src/ATen/native/cuda/ReplicationPadding.cu:136]: We can change those `channel`, `height`, `width` to be corresponding tensor strides and slightly modify the `idx_2d` function. How does that sound?

Otherwise, we may have to add `gradInput.contiguous()` which may affect performance.
